src/lib/logstore/log_dev.cpp

```diff
@@ -105,26 +108,17 @@ void LogDev::start(bool format) {
 }
 
 void LogDev::stop() {
-    THIS_LOGDEV_LOG(INFO, "Logdev stopping log_dev={}", m_logdev_id);
-    HS_LOG_ASSERT((m_pending_flush_size == 0), "LogDev stop attempted while writes to logdev are pending completion");
-    const bool locked_now = run_under_flush_lock([this]() {
-        {
-            std::unique_lock< std::mutex > lk{m_block_flush_q_mutex};
-            m_stopped = true;
-        }
-        m_block_flush_q_cv.notify_one();
-        return true;
-    });
-
-    if (!locked_now) { THIS_LOGDEV_LOG(INFO, "LogDev stop is queued because of pending flush or truncation ongoing"); }
-
+    THIS_LOGDEV_LOG(INFO, "Logdev stopping id {}", m_logdev_id);
+    HS_LOG_ASSERT((m_pending_flush_size.load() == 0),
+                  "LogDev stop attempted while writes to logdev are pending completion");
     {
-        // Wait for the stopped to be completed
-        std::unique_lock< std::mutex > lk{m_block_flush_q_mutex};
-        m_block_flush_q_cv.wait(lk, [&] { return m_stopped; });
+        std::unique_lock lg = flush_guard();
+        m_stopped = true;
     }
+    // after we call stop, we need to do any pending device truncations
+    truncate();
```

Comment:
sometimes , log store truncation will only do truncation in memory, they will not issue log dev truncation.
the idea here is  when we stopping log dev, doing a truncation here can remove those stale log as much as possible, which is only be truncated in memory. generally , on log dev can sustain multiple log store, this will make the data which needs to be loaded as less as possible.

logdev truncation is a  best-effort  truncation. that means if raft and  resource manager does  not issue any truncation(log store truncation), log dev truncation will do nothing. log dev truncation  will only truncate those which has already been truncated by log store.